### PR TITLE
refactor!: Add `BigDecimalExt` and remove `IntermediateDecimal`

### PR DESCRIPTION
**Pull request by Bhavyajain21**
/claim #113

### CODE REVIEW

**Comment by Bhavyajain21** [comment]
@JayWhite2357 please review!

**Review by iajoiner — COMMENTED** [review]
@Bhavyajain21 Really thanks for your hard work!

I don't think a third crate is really helpful here. Instead what would be great is getting the `proof-of-sql` bigdecimals into the `proof-of-sql-parser` crate.

**Comment by Bhavyajain21** [comment]
Hey @iajoiner , will it not work with this piece of code?

**Comment by iajoiner** [comment]
@Bhavyajain21 Well I think your work is pretty good. Please do not add another crate though.

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,12555 @@
+{
```
Comment (iajoiner) [inline package-lock.json:1]: Please remove this file and add it to `.gitignore`.

Comment (Bhavyajain21) [reply package-lock.json]: Done!

**Comment by Bhavyajain21** [comment]
> @Bhavyajain21 Well I think your work is pretty good. Please do not add another crate though. Please read #113 again carefully.

Done, can you please review now? @iajoiner

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [workspace]
 resolver = "2"
-members = ["crates/proof-of-sql", "crates/proof-of-sql-parser"]
+members = ["crates/proof-of-sql", "crates/proof-of-sql-parser", "crates/proof-of-sql-utils"]
```
Comment (iajoiner) [inline Cargo.toml:3]: @Bhavyajain21 The crate no longer exists so please remove it here as well

Comment (Bhavyajain21) [reply Cargo.toml:3]: On it

Comment (Bhavyajain21) [reply Cargo.toml]: Done! @iajoiner

File: crates/proof-of-sql-parser/src/big_decimal_ext.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    OutOfRange,
+    #[snafu(display("Fractional part of decimal is non-zero"))]
+    LossyCast,
+}
+
+impl BigDecimalExt for BigDecimal {
+    /// Calculate precision after trimming leading and trailing zeros
+    fn precision(&self) -> u8 {
+        let trimmed = self.normalized();
+        trimmed.digits() as u8
```
Comment (iajoiner) [inline crates/proof-of-sql-parser/src/big_decimal_ext.rs:27]:
```rust
self.normalized().digits() as u8
```

Comment (Bhavyajain21) [reply crates/proof-of-sql-parser/src/big_decimal_ext.rs]: Done!

File: crates/proof-of-sql-parser/src/big_decimal_ext.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    /// Calculate scale (fractional digits) after trimming trailing zeros
+    fn scale(&self) -> i8 {
+        let trimmed = self.normalized();
+        trimmed.fractional_digit_count() as i8
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply crates/proof-of-sql-parser/src/big_decimal_ext.rs]: Done!

**Review by iajoiner — COMMENTED** [review]
@Bhavyajain21 Really thanks for adding `BigDecimalExt` and eliminating `IntermediateDecimal`. Let's remove `proof_of_sql::base::math::decimal::Decimal` as well maybe in a future PR. :)

Note that the bounty will only be awarded once all the requested changes are complete.

**Comment by iajoiner** [comment]
@Bhavyajain21 Please get CI to pass. Also please pay attention to how I edited your PR title. For future references please stick to our PR guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

**Comment by iajoiner** [comment]
> I noticed that the removal of proof_of_sql::base::math::decimal::Decimal is mentioned as an optional task in the issue description. While I’m happy to raise a follow-up PR for that, I believe the bounty should still be awarded based on the completion of the core requirements of this issue, as the optional change shouldn’t affect the overall reward eligibility.
@Bhavyajain21 Not really. The optional changes are
- [ ] It is not clear if the DecimalError enum is needed or if bigdecimal::ParseBigDecimalError or similar is sufficient.
- [ ] The scale_scalar function should likely be moved into the proof_of_sql::base::scalar module.
Let's get this one to pass and then remove `Decimal` in the next PR. We need @JayWhite2357 for final approvals (see Check Approver in CI) but otherwise if you can get the rest of CI to pass I will approve.

File: crates/proof-of-sql-parser/src/big_decimal_ext.rs [hunk 5]
```diff
@@ -0,0 +1,53 @@
+use bigdecimal::BigDecimal; // Import BigDecimal
+use num_bigint::{BigInt, ToBigInt}; // Import BigInt and ToBigInt
+use snafu::Snafu;
```
[the remark this answers is not in the document]
Comment (Bhavyajain21) [reply crates/proof-of-sql-parser/src/big_decimal_ext.rs]: Done

**Comment by JayWhite2357** [comment]
Closing in favor of https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/283. See https://github.com/spaceandtimelabs/sxt-proof-of-sql/issues/113#issuecomment-2424402223.

**Pull request closed without merging**